src/foam/core/JSON.js

```diff
@@ -382,11 +381,12 @@ foam.CLASS({
     },
 
     function outputDate(o) {
-      if ( this.formatDatesAsNumbers ) {
-        this.out(o.valueOf());
-      } else {
-        this.out(JSON.stringify(o));
-      }
+      // Serialize an unambiguous timestamp.  Date and DateTime
+      // property can provide an alternative toJSON() and
+      // adapt/javaJSONParser mechanism to save space.
+      this.out('{"class":"__Timestamp__","value":');
```

Comment:
I disagree with this approach for two reasons:
1. It is non standard and will seem odd and unusable by non-FOAM users.
2. It will be slow because it will use extra bandwidth and CPU when communicating between FOAM instances, but between FOAM instances we should be able to encode it however we like to maximize performance.

Reply:
What do you suggest we do instead? There's no real standard for dates in json so we're trying to invent our own here.
For models with date properties, we can provide alternate toJSON and adapt/javaJSONParser to maximize performance so I don't think we lose much here with FOAM to FOAM communication.

Reply:
> It is non standard and will seem odd and unusable by non-FOAM users.

So is __Class__ and __Property__.  There is no standard way of serializing Dates to JSON.

> It will be slow because it will use extra bandwidth and CPU when communicating between FOAM instances, but between FOAM instances we should be able to encode it however we like to maximize performance.

Only when sending a Date not as part of a Date property.  What alternative do we have?  The previous strategy of just serializing a Date as a number means that we lose the type information and we end up having difficult to debug errors where method arguments come in as Longs or Ints instead of Dates.

In practice we won't use much more bandwidth because any time we're using a Date property, it has custom serialization to send just the date as a number.  But in cases where we're not a Date property, like an Array, method arguments, some value of a Map, then serialization will still work.

Reply:
Another thing we could do is entirely remove support from foam.core.JSON for outputting things that don't exist in JSON (Dates, Classes, Properties. Enums).  And to serialize those types require support to be added to the appropriate Properties and Arguments.

This may provide a cleaner separation of concerns and easier understanding.  JSON Outputter doesn't know about (and can error on) if you try to output any value that doesn't exist in JSON. And any "place" that can store these other non-JSON types can translate them to whatever JSON primitive they deem appropriate.  A plain number or string for dates, an ID string for classes, a plain ordinal for enums (optionally with a type if necessary).